src/uu/sed/src/compiler.rs

```diff
@@ -202,31 +196,177 @@ fn build_command_map() -> HashMap<char, CommandSpec> {
     formats.into_iter().map(|f| (f.code, f)).collect()
 }
 
+/// Compile the scripts into an executable data structure.
 pub fn compile(
     scripts: Vec<ScriptValue>,
-    processing_context: &mut ProcessingContext,
+    context: &mut ProcessingContext,
 ) -> UResult<Option<Rc<RefCell<Command>>>> {
     let mut make_providers = ScriptLineProvider::new(scripts);
 
     let mut empty_line = ScriptCharProvider::new("");
-    let result = compile_thread(&mut make_providers, &mut empty_line, processing_context)?;
-    // TODO: fix-up labels, check used labels, setup append & match structures
+    let result = compile_sequence(&mut make_providers, &mut empty_line, context)?;
+
+    // Link the ends of command blocks to their following commands.
+    if context.parsed_block_nesting > 0 {
+        return Err(USimpleError::new(1, "unmatched `{'"));
+    }
+    patch_block_endings(result.clone());
+
+    // Link branch commands to the target label commands.
+    populate_label_map(result.clone(), context)?;
+    resolve_branch_targets(result.clone(), context)?;
+
+    // Comment-out the following to show the compiled script.
+    #[cfg(any())]
+    dbg!(&result);
+
+    // TODO: setup append & match structures
     Ok(result)
 }
 
-// Compile provided scripts into a thread of commands
-fn compile_thread(
+/// For every Command in the top-level `head` chain, look for
+/// `CommandData::Block(Some(sub_head))`.  Recursively patch
+/// the sub-chain, then splice its tail back to the original
+/// “next” pointer of the *parent* (falling back to its own
+/// parent_next if its own next was `None`).
+fn patch_block_endings(head: Option<Rc<RefCell<Command>>>) {
```

Comment:
would it make sense to write unit test for this function ?

Reply:
I have; search for "// patch_block_endings".

Reply:
oh, sorry